<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Changelog • snpEnrichment</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/darkly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Changelog" />


<meta name="robots" content="noindex">

<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">snpEnrichment</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="In-development version">1.8.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/mcanouil/snpEnrichment/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      <small>Source: <a href='https://github.com/mcanouil/snpEnrichment/blob/master/NEWS.md'><code>NEWS.md</code></a></small>
    </div>

    <div id="snpenrichment-180" class="section level1">
<h1 class="page-header" data-toc-text="1.8.0">
<a href="#snpenrichment-180" class="anchor"></a>snpEnrichment 1.8.0<small> Unreleased </small>
</h1>
<div id="minor-improvements-and-fixes" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-improvements-and-fixes" class="anchor"></a>Minor improvements and fixes</h2>
<ul>
<li>In <code>R/*estimate_ethnicity*.R</code>,
<ul>
<li>use tidy style.</li>
</ul>
</li>
<li>In <code>DESCRIPTION</code>,
<ul>
<li>a little change in Title and Description to clearly state the package uses a sampling approach.</li>
<li>setup package to use roxygen2.</li>
</ul>
</li>
</ul>
</div>
</div>
    <div id="snpenrichment-170" class="section level1">
<h1 class="page-header" data-toc-text="1.7.0">
<a href="#snpenrichment-170" class="anchor"></a>snpEnrichment 1.7.0<small> 2015-10-01 </small>
</h1>
<ul>
<li>NAMESPACE: Update to fix warnings from R CMD check using R version 3.2.2.</li>
<li>DESCRIPTION: Update according to new imports in NAMESPACE.</li>
</ul>
</div>
    <div id="snpenrichment-166" class="section level1">
<h1 class="page-header" data-toc-text="1.6.6">
<a href="#snpenrichment-166" class="anchor"></a>snpEnrichment 1.6.6<small> Unreleased </small>
</h1>
<ul>
<li>NAMESPACE: Update to fix warnings from R CMD check using R version 3.2.2.</li>
<li>R/enrichment.R: Update ‘plot’ method to fix global definition warnings.</li>
</ul>
</div>
    <div id="snpenrichment-165" class="section level1">
<h1 class="page-header" data-toc-text="1.6.5">
<a href="#snpenrichment-165" class="anchor"></a>snpEnrichment 1.6.5<small> Unreleased </small>
</h1>
<ul>
<li>tests/RUN-TEST_snpEnrichment.R: Add a test script.</li>
</ul>
</div>
    <div id="snpenrichment-164" class="section level1">
<h1 class="page-header" data-toc-text="1.6.4">
<a href="#snpenrichment-164" class="anchor"></a>snpEnrichment 1.6.4<small> Unreleased </small>
</h1>
<ul>
<li>R/enrichment.R: Update ‘getEnrichSNP’ to add information about related eSNP for xSNP.</li>
</ul>
</div>
    <div id="snpenrichment-163" class="section level1">
<h1 class="page-header" data-toc-text="1.6.3">
<a href="#snpenrichment-163" class="anchor"></a>snpEnrichment 1.6.3<small> 2015-02-09 </small>
</h1>
<ul>
<li>DESCRIPTION: Change maintainer email.</li>
<li>R/enrichment.R: Minor changes.</li>
</ul>
</div>
    <div id="snpenrichment-162" class="section level1">
<h1 class="page-header" data-toc-text="1.6.2">
<a href="#snpenrichment-162" class="anchor"></a>snpEnrichment 1.6.2<small> Unreleased </small>
</h1>
<ul>
<li>R/enrichment.R: Minor changes.</li>
<li>R/chromosome.R: Minor changes.</li>
<li>R/Global.R: Minor changes.</li>
</ul>
</div>
    <div id="snpenrichment-161" class="section level1">
<h1 class="page-header" data-toc-text="1.6.1">
<a href="#snpenrichment-161" class="anchor"></a>snpEnrichment 1.6.1<small> Unreleased </small>
</h1>
<ul>
<li>R/enrichment.R: Minor changes.</li>
</ul>
</div>
    <div id="snpenrichment-160" class="section level1">
<h1 class="page-header" data-toc-text="1.6.0">
<a href="#snpenrichment-160" class="anchor"></a>snpEnrichment 1.6.0<small> 2014-11-04 </small>
</h1>
<ul>
<li>R/Global.R: Fix a bug in ‘compareEnrichment’ method with parameter ‘empiricPvalue=TRUE’.</li>
</ul>
</div>
    <div id="snpenrichment-1510" class="section level1">
<h1 class="page-header" data-toc-text="1.5.10">
<a href="#snpenrichment-1510" class="anchor"></a>snpEnrichment 1.5.10<small> Unreleased </small>
</h1>
<ul>
<li>R/Global.R: Fix a glitch in ‘readEnrichment’. Sometimes chromosomes 10 and 20 are read respectively instead of 1 and 2.</li>
<li>R/Global.R: Some minor changes.</li>
</ul>
</div>
    <div id="snpenrichment-159" class="section level1">
<h1 class="page-header" data-toc-text="1.5.9">
<a href="#snpenrichment-159" class="anchor"></a>snpEnrichment 1.5.9<small> Unreleased </small>
</h1>
<ul>
<li>data/toyEnrichment.RData: Minor changes.</li>
<li>R/Global.R: Add internal function ‘.splitByChrom’. User is now allowed to use a single file in ‘snpListDir’ for ‘readEnrichment’ function.</li>
<li>man/snpEnrichment-internal.Rd: Update according to R code changes.</li>
<li>man/readEnrichment.Rd: Update according to R code changes.</li>
</ul>
</div>
    <div id="snpenrichment-158" class="section level1">
<h1 class="page-header" data-toc-text="1.5.8">
<a href="#snpenrichment-158" class="anchor"></a>snpEnrichment 1.5.8<small> Unreleased </small>
</h1>
<ul>
<li>R/enrichment.R: Update ‘print’ method to print results according to the parameter ‘empiricPvalue’ in ‘reSample’ function.</li>
<li>R/enrichment.R: Update ‘compareEnrichment’ according to changes in ‘print’ method.</li>
<li>R/chromosome.R: Some minor changes.</li>
<li>man/compareEnrichment.Rd: Update according to R code changes.</li>
<li>man/print-methods.Rd: Update according to R code changes.</li>
</ul>
</div>
    <div id="snpenrichment-157" class="section level1">
<h1 class="page-header" data-toc-text="1.5.7">
<a href="#snpenrichment-157" class="anchor"></a>snpEnrichment 1.5.7<small> Unreleased </small>
</h1>
<ul>
<li>R/*.R: Add function ‘getEnrichSNP’ and some minor corrections.</li>
<li>man/*.Rd: Update according to R code changes and some corrections.</li>
<li>NAMESPACE: Updata according to R code changes.</li>
<li>inst/example.R: Update ‘excludeFile’ list according to changes in ‘toyEnrichment’ and ‘extdata’.</li>
<li>man/snpEnrichment-package.Rd: Update ‘excludeFile’ list according to changes in ‘toyEnrichment’ and ‘extdata’.</li>
<li>man/excludeSNP-methods.Rd: Update ‘excludeFile’ list according to changes in ‘toyEnrichment’ and ‘extdata’.</li>
<li>README.md: Update ‘excludeFile’ list according to changes in ‘toyEnrichment’ and ‘extdata’.</li>
</ul>
</div>
    <div id="snpenrichment-156" class="section level1">
<h1 class="page-header" data-toc-text="1.5.6">
<a href="#snpenrichment-156" class="anchor"></a>snpEnrichment 1.5.6<small> Unreleased </small>
</h1>
<ul>
<li>R/Global.R: ‘readEnrichment’ works with at least one snp list in ‘snpListDir’.</li>
<li>R/Global.R: Fix ‘mclapply2’ under operating systems which are not Linux.</li>
<li>R/Global.R: Fix ‘maxCores’ under operating systems which are not Linux.</li>
<li>man/snpEnrichment-package.Rd: Update example.</li>
<li>inst/example.R: Update according to changes in ‘snpEnrichment-package.Rd’.</li>
<li>README.md: Update according to changes in ‘snpEnrichment-package.Rd’.</li>
<li>R/*.R: the “types” parameter is renamed “type”.</li>
<li>man/*.Rd: Update according to R code changes.</li>
</ul>
</div>
    <div id="snpenrichment-155" class="section level1">
<h1 class="page-header" data-toc-text="1.5.5">
<a href="#snpenrichment-155" class="anchor"></a>snpEnrichment 1.5.5<small> Unreleased </small>
</h1>
<ul>
<li>man/*.Rd: Replace ‘summary’ by ‘print’.</li>
<li>inst/example.R: Replace ‘summary’ by ‘print’ according to changes in ’*.Rd’ files.</li>
<li>README.md: Replace ‘summary’ by ‘print’ according to changes in ’*.Rd’ files.</li>
</ul>
</div>
    <div id="snpenrichment-154" class="section level1">
<h1 class="page-header" data-toc-text="1.5.4">
<a href="#snpenrichment-154" class="anchor"></a>snpEnrichment 1.5.4<small> Unreleased </small>
</h1>
<ul>
<li>R/*.R: Change default value for ‘empiricPvalue’ parameter to ‘TRUE’.</li>
<li>man/*.Rd: Update according to R code changes.</li>
<li>inst/example.R: Update according to R code changes.</li>
<li>README.md: Update according to R code changes.</li>
<li>R/enrichment.R: Change warning messages in ‘compareEnrichment’.</li>
</ul>
</div>
    <div id="snpenrichment-152" class="section level1">
<h1 class="page-header" data-toc-text="1.5.2">
<a href="#snpenrichment-152" class="anchor"></a>snpEnrichment 1.5.2<small> Unreleased </small>
</h1>
<ul>
<li>R/Global.R: Add some internal functions to check files needed by ‘initFiles’, ‘writeLD’ and ‘readEnrichment’ functions.</li>
<li>man/snpEnrichment-internal.Rd: Update according to R code changes.</li>
<li>inst/example.R: Update according to R code changes.</li>
</ul>
</div>
    <div id="snpenrichment-151" class="section level1">
<h1 class="page-header" data-toc-text="1.5.1">
<a href="#snpenrichment-151" class="anchor"></a>snpEnrichment 1.5.1<small> Unreleased </small>
</h1>
<ul>
<li>R/Global.R: Internal function ‘maxCores’ handle warnings on macintosh system.</li>
<li>R/Global.R: Handle warning / error in ‘initFiles’ when ‘snpInfoDir’ or ‘signalFile’ are missing or invalid.</li>
</ul>
</div>
    <div id="snpenrichment-150" class="section level1">
<h1 class="page-header" data-toc-text="1.5.0">
<a href="#snpenrichment-150" class="anchor"></a>snpEnrichment 1.5.0<small> Unreleased </small>
</h1>
<ul>
<li>R/Global.R: Fix a bug in ‘compareEnich’ when ‘xSNP’ was missing in "object.*".</li>
<li>man/*.Rd: Some corrections in example.</li>
</ul>
</div>
    <div id="snpenrichment-143" class="section level1">
<h1 class="page-header" data-toc-text="1.4.3">
<a href="#snpenrichment-143" class="anchor"></a>snpEnrichment 1.4.3<small> Unreleased </small>
</h1>
<ul>
<li>man/*.Rd: Update according to R code changes.</li>
</ul>
</div>
    <div id="snpenrichment-142" class="section level1">
<h1 class="page-header" data-toc-text="1.4.2">
<a href="#snpenrichment-142" class="anchor"></a>snpEnrichment 1.4.2<small> Unreleased </small>
</h1>
<ul>
<li>R/enrichment.R: Fix a bug when ‘xSNP’ field is empty in ‘plot’ method.</li>
<li>R/Global.R: Fix ‘options(stringsAsFactors=FALSE)’ in ‘writeLD’.</li>
<li>man/snpEnrichment-package.Rd: Update according to DESCRIPTION file.</li>
</ul>
</div>
    <div id="snpenrichment-141" class="section level1">
<h1 class="page-header" data-toc-text="1.4.1">
<a href="#snpenrichment-141" class="anchor"></a>snpEnrichment 1.4.1<small> Unreleased </small>
</h1>
<ul>
<li>R/*.R: Minor optimizations.</li>
<li>R/enrichment.R: Fix a bug in ‘plot’ method with ‘ggplot=TRUE’.</li>
</ul>
</div>
    <div id="snpenrichment-140" class="section level1">
<h1 class="page-header" data-toc-text="1.4.0">
<a href="#snpenrichment-140" class="anchor"></a>snpEnrichment 1.4.0<small> Unreleased </small>
</h1>
<ul>
<li>R/Global.R: Change internal functions used by ‘readEnrichment’ and ‘writeLD’.</li>
<li>R/Global.R: Improvment in ‘writeLD’ function.</li>
<li>R/Global.R: Add ‘depth’ parameter in ‘writeLD’ function.</li>
<li>R/Global.R: Minor optimizations.</li>
<li>R/enrichment.R: Fix a bug when ‘object’ was not updated properly, with ‘reSample’ function and all defaults parameters.</li>
<li>R/*.R: Remove ‘extendMethod’ parameters from all functions.</li>
<li>data/: Remove ‘toyEnrichmentM2.RData’.</li>
<li>data/: Rename ‘toyEnrichmentM1.RData’ to ‘toyEnrichment.RData’</li>
<li>man/*.Rd: Update according to R code changes.</li>
</ul>
</div>
    <div id="snpenrichment-1312" class="section level1">
<h1 class="page-header" data-toc-text="1.3.12">
<a href="#snpenrichment-1312" class="anchor"></a>snpEnrichment 1.3.12<small> Unreleased </small>
</h1>
<ul>
<li>R/*.R: Minor changes in R code.</li>
<li>R/Global.R: Remove parameter ‘snpListDir’ in ‘initFiles’ function.</li>
<li>R/Global.R: Fix ‘ldThresh’ parameter in ‘initFiles’ function.</li>
<li>man/*.Rd: Update according to R code changes.</li>
</ul>
</div>
    <div id="snpenrichment-1311" class="section level1">
<h1 class="page-header" data-toc-text="1.3.11">
<a href="#snpenrichment-1311" class="anchor"></a>snpEnrichment 1.3.11<small> Unreleased </small>
</h1>
<ul>
<li>R/enrichment.R: Add ‘pvalue’ parameters in ‘plot’ method.</li>
<li>R/plot-methods.Rd: Update according to R code changes.</li>
</ul>
</div>
    <div id="snpenrichment-1310" class="section level1">
<h1 class="page-header" data-toc-text="1.3.10">
<a href="#snpenrichment-1310" class="anchor"></a>snpEnrichment 1.3.10<small> Unreleased </small>
</h1>
<ul>
<li>R/*.R: Add ‘empiricPvalue’ arguments in order to choose if p-values should be empirical or based on null distribution (resampling).</li>
</ul>
</div>
    <div id="snpenrichment-139" class="section level1">
<h1 class="page-header" data-toc-text="1.3.9">
<a href="#snpenrichment-139" class="anchor"></a>snpEnrichment 1.3.9<small> Unreleased </small>
</h1>
<ul>
<li>R/Enrichment.R: Prevent a bug in ‘plot’ method with ‘ggplot=TRUE’.</li>
</ul>
</div>
    <div id="snpenrichment-138" class="section level1">
<h1 class="page-header" data-toc-text="1.3.8">
<a href="#snpenrichment-138" class="anchor"></a>snpEnrichment 1.3.8<small> Unreleased </small>
</h1>
<ul>
<li>R/*.R: Minor changes in R code.</li>
<li>R/Global.R: Fix a missing square root in Z statistics computation.</li>
<li>R/enrichment.R: Fix legend position in ‘plot’ for ‘ggplot=TRUE’.</li>
</ul>
</div>
    <div id="snpenrichment-137" class="section level1">
<h1 class="page-header" data-toc-text="1.3.7">
<a href="#snpenrichment-137" class="anchor"></a>snpEnrichment 1.3.7<small> Unreleased </small>
</h1>
<ul>
<li>R/*.R: Minor changes in R code.</li>
<li>R/Global.R: Change p-value computation in ‘.reSample’ and ‘compareEnrich’ functions.</li>
<li>R/Global.R: Convergence control on Z statistic is removed in ‘.reSample’ and ‘compareEnrich’ functions.</li>
</ul>
</div>
    <div id="snpenrichment-136" class="section level1">
<h1 class="page-header" data-toc-text="1.3.6">
<a href="#snpenrichment-136" class="anchor"></a>snpEnrichment 1.3.6<small> Unreleased </small>
</h1>
<ul>
<li>R/*.R: Minor changes in R code.</li>
<li>R/enrichment.R: ‘compareEnrichment’ now returns ‘object.x’, ‘object.y’ and object from comparison.</li>
<li>R/compareEnrichment.Rd: Update according to R code changes.</li>
</ul>
</div>
    <div id="snpenrichment-135" class="section level1">
<h1 class="page-header" data-toc-text="1.3.5">
<a href="#snpenrichment-135" class="anchor"></a>snpEnrichment 1.3.5<small> Unreleased </small>
</h1>
<ul>
<li>R/*.R: Minor changes in R code.</li>
<li>R/enrichment.R: Fix an error in ‘reSample’ method that occurs in ‘compareEnrichment’ method.</li>
<li>R/enrichment.R: Add controls on installed packages for ‘ggplot=TRUE’ in plot method.</li>
<li>R/enrichment.R: Remove convergence control in ‘reSample’ when ‘Resampling’ slot are empty in ‘compareEnrichment’. ‘compareEnrichment’ now performs ‘nSample’ resampling for ‘object.x’, ‘object.y’ and for comparison.</li>
<li>R/global.R: Remove ‘while’ loop in ‘.compareEnrich’ (not used yet).</li>
</ul>
</div>
    <div id="snpenrichment-134" class="section level1">
<h1 class="page-header" data-toc-text="1.3.4">
<a href="#snpenrichment-134" class="anchor"></a>snpEnrichment 1.3.4<small> Unreleased </small>
</h1>
<ul>
<li>R/*.R: Minor changes in R code.</li>
<li>man/compareEnrichment.Rd: Add alias for signature “Enrichment”.</li>
</ul>
</div>
    <div id="snpenrichment-133" class="section level1">
<h1 class="page-header" data-toc-text="1.3.3">
<a href="#snpenrichment-133" class="anchor"></a>snpEnrichment 1.3.3<small> Unreleased </small>
</h1>
<ul>
<li>R/enrichment.R: Fix a bug in ‘excludeSNP’ when ‘LD’ is set to ‘FALSE’.</li>
</ul>
</div>
    <div id="snpenrichment-132" class="section level1">
<h1 class="page-header" data-toc-text="1.3.2">
<a href="#snpenrichment-132" class="anchor"></a>snpEnrichment 1.3.2<small> Unreleased </small>
</h1>
<ul>
<li>R/enrichment.R: Redefine ‘print’ method for ‘Enrichment’ object.</li>
<li>R/chromosome.R: Redefine ‘print’ method for ‘Chromosome’ object.</li>
</ul>
</div>
    <div id="snpenrichment-131" class="section level1">
<h1 class="page-header" data-toc-text="1.3.1">
<a href="#snpenrichment-131" class="anchor"></a>snpEnrichment 1.3.1<small> Unreleased </small>
</h1>
<ul>
<li>R/Global.R: ‘readEnrichment’ returns an error when ‘initFiles’ was not run before.</li>
<li>R/enrichment.R: ‘reSample’ method for ‘Enrichment’ object now reruns one warning when ‘mc.cores’ is decreased.</li>
<li>R/enrichment.R: call for ‘reSample’ method is saved properly in ‘Enrichment’ object.</li>
<li>R/enrichment.R: Redefine ‘plot’ method for ‘Enrichment’ object. ‘plot’ method can use ggplot2, if ‘ggplot’ is set to ‘TRUE’.</li>
<li>man/*.Rd: Update documentations files according to R code modification.</li>
</ul>
</div>
    <div id="snpenrichment-130" class="section level1">
<h1 class="page-header" data-toc-text="1.3.0">
<a href="#snpenrichment-130" class="anchor"></a>snpEnrichment 1.3.0<small> Unreleased </small>
</h1>
<ul>
<li>R/Global.R: ‘readEnrichment’ can deal with ld files computed with plink directly.</li>
</ul>
</div>
    <div id="snpenrichment-123" class="section level1">
<h1 class="page-header" data-toc-text="1.2.3">
<a href="#snpenrichment-123" class="anchor"></a>snpEnrichment 1.2.3<small> Unreleased </small>
</h1>
<ul>
<li>man/EnrichSNP-class.Rd: Add alias for method ‘[’ and ‘[’ with signature = ‘EnrichSNP,ANY,ANY,ANY’.</li>
<li>man/Chromosome-class.Rd: Add alias for method ‘[’ and ‘[’ with signature = ‘Chromosome,ANY,ANY,ANY’.</li>
<li>man/Enrichment-class.Rd: Add alias for method ‘[’ and ‘[’ with signature = ‘Enrichment,ANY,ANY,ANY’.</li>
<li>R/Global.R: Now use ‘read.plink’, ‘col.summary’, ‘ld’ from ‘snpStats’ package for read plink files and compute MAF and LD.</li>
<li>R/Enrichment.R: Fix ‘show’ method in order to deal with new parameters in ‘readEnrichment’ and ‘reSample’ functions.</li>
<li>R/*.R: Change default setting for ‘mc.cores’ to 1.</li>
<li>man/*.Rd: Update documentations files according to R code modification.</li>
</ul>
</div>
    <div id="snpenrichment-122" class="section level1">
<h1 class="page-header" data-toc-text="1.2.2">
<a href="#snpenrichment-122" class="anchor"></a>snpEnrichment 1.2.2<small> Unreleased </small>
</h1>
<ul>
<li>R/*.R: Change name of ‘SNP’ class to ‘EnrichSNP’.</li>
<li>R/*.R: Some improvements and computation time optimisations.</li>
<li>R/Global.R: Add ‘ldDir’ parameter in ‘writeLD’ which allow to choose the directory where LD files should be written.</li>
<li>R/Global.R: Add ‘ldDir’ parameter in ‘readEnrichment’.</li>
<li>R/Global.R: ‘initFiles’ now write and clean tempory files used by ‘readEnrichment’.</li>
<li>R/Global.R: Now LD is only computed by ‘writeLD’. Not anymore by ‘initFiles’.</li>
<li>R/Global.R: Remove ‘ldThresh’ and ‘LD’ parameters from ‘initFiles’.</li>
<li>R/Global.R: Remove ‘signalFile’, ‘transcriptFile’, ‘snpInfoDir’, ‘distThresh’, ‘LD’, ‘ldThresh’, ‘sigThresh’, ‘MAFpool’ and ‘extendMethod’ parameters from ‘compareEnrichment’.</li>
<li>R/enrichment.R: Remove ‘signalFile’, ‘transcriptFile’, ‘snpInfoDir’, ‘distThresh’, ‘LD’, ‘ldThresh’, ‘sigThresh’, ‘MAFpool’ and ‘extendMethod’ parameters from ‘compareEnrichment’.</li>
<li>R/enrichment.R: Fix issue for the whole genome convergence plot in ‘plot’ methods. Add parameter ‘onlyGenome’.</li>
<li>man/*.Rd: Update documentations files according to R code modification.</li>
</ul>
</div>
    <div id="snpenrichment-121" class="section level1">
<h1 class="page-header" data-toc-text="1.2.1">
<a href="#snpenrichment-121" class="anchor"></a>snpEnrichment 1.2.1<small> Unreleased </small>
</h1>
<ul>
<li>R/*-class.R: Add ‘print’ methods (same as ‘show’).</li>
<li>R/Global.R: Add error messages when files can not be created by ‘initFiles’.</li>
<li>R/Global.R: Remove ‘require(parallel)’ from ‘mclapply2’ function.</li>
<li>exdata/*: Move all external data to ‘inst’ directory in order to be used in examples.</li>
<li>man/*.Rd: Update documentations files according to R code modification.</li>
</ul>
</div>
    <div id="snpenrichment-120" class="section level1">
<h1 class="page-header" data-toc-text="1.2.0">
<a href="#snpenrichment-120" class="anchor"></a>snpEnrichment 1.2.0<small> Unreleased </small>
</h1>
<ul>
<li>R/enrichment.R: Add ‘onlyGenome’ parameter, which allow to compute the resampling step at chromosome level and genome or only at genome level.</li>
<li>R/enrichment.R: Fix ‘plot’ methods for low number of resampling.</li>
<li>R/enrichment.R: Change ‘Parameters’ to “Call”.</li>
<li>R/enrichment.R: ‘show’ is available for ‘Call’ slot.</li>
<li>R/enrichment.R: Remove ‘sigThresh’ and ‘extendMethod’ parameters from “reSample” method with ‘signature=“Enrichment”’.</li>
<li>R/enrichment.R: Remove ‘sigThresh’ and ‘extendMethod’ parameters from “excludeSNP” methods.</li>
<li>R/enrichment.R: Fix abscisse according to ‘nSample’. Abscisse origin is now equal to three.</li>
<li>R/enrichment.R: ‘chrNumber’ is replaced by ‘chr’, in ‘plot’ method for ‘Enrichment’ object.</li>
<li>R/Global.R: Fix bug in ‘compareEnrichment’ when some fields were not reseted with the option ‘onlyGenome=TRUE’.</li>
<li>R/Global.R: Remove ‘onlySignal’ parameter from ‘writeLD’ function.</li>
<li>R/Global.R: Change ‘Parameters’ to “Call”.</li>
<li>R/Global.R: Remove ‘sigThresh’ and ‘extendMethod’ parameters from “excludeSNP” methods.</li>
<li>NAMESPACE: Fix ‘NAMESPACE’ with missing ‘SNP’ class and methods.</li>
<li>data/*: Generate new toy datasets.</li>
<li>extdata/*: Generate new external datasets (iles needed to run the examples for the packages functions).</li>
<li>man/*.Rd: Update documentations files according to R code modification.</li>
</ul>
</div>
    <div id="snpenrichment-114" class="section level1">
<h1 class="page-header" data-toc-text="1.1.4">
<a href="#snpenrichment-114" class="anchor"></a>snpEnrichment 1.1.4<small> Unreleased </small>
</h1>
<ul>
<li>R/enrichment.R: Now results from method ‘compareEnrichment’ are the enrichment object and the summary table.</li>
<li>R/enrichment.R: ‘nSample’ in ‘excludeSNP’ can be equal to 0.</li>
<li>R/Global.R: ‘maxCores’ return at least 1.</li>
<li>R/mclapply2.R: run in parallel mode with at least 1 core.</li>
</ul>
</div>
    <div id="snpenrichment-113" class="section level1">
<h1 class="page-header" data-toc-text="1.1.3">
<a href="#snpenrichment-113" class="anchor"></a>snpEnrichment 1.1.3<small> Unreleased </small>
</h1>
<ul>
<li>R/Global.R: Some improvements and changes of functions: ‘initFiles’, ‘.writeSignal’, ‘writeLD’.</li>
</ul>
</div>
    <div id="snpenrichment-112" class="section level1">
<h1 class="page-header" data-toc-text="1.1.2">
<a href="#snpenrichment-112" class="anchor"></a>snpEnrichment 1.1.2<small> Unreleased </small>
</h1>
<ul>
<li>R/Global.R: Add informations about snp losses on each step of the algorithm.</li>
<li>R/Global.R: Fix minor bugs in read functions.</li>
</ul>
</div>
    <div id="snpenrichment-111" class="section level1">
<h1 class="page-header" data-toc-text="1.1.1">
<a href="#snpenrichment-111" class="anchor"></a>snpEnrichment 1.1.1<small> Unreleased </small>
</h1>
<ul>
<li>man/plot-methods.Rd: Corrects the description of method plot.</li>
</ul>
</div>
    <div id="snpenrichment-110" class="section level1">
<h1 class="page-header" data-toc-text="1.1.0">
<a href="#snpenrichment-110" class="anchor"></a>snpEnrichment 1.1.0<small> Unreleased </small>
</h1>
<ul>
<li>R/Global.R: Some minor code optimization and simplification.</li>
<li>R/Global.R: The method ‘excludeSNP’ now exclude the extended list of SNP provided in ‘excludeFile’. Note: ‘excludeFile’ can be a character vector with SNPs or a file path which contains a SNPs list.</li>
<li>R/Global.R: The function ‘writeLD’ now compute Linkage Disequilibrium for all SNPs in reference data or only for SNPs in signal.</li>
<li>R/enrichment.R: The slot ‘Signal’ of Enrichment object is replaced by slot ‘Loss’. This slot gives information about data losses (e.g. SNPs removed).</li>
<li>R/Global.R: Add ‘plot’ method for Enrichment object. This method show the convergence of Z statistics. Z statistic is scaled and centered in order to compare the convergence in chromosomes/genome.</li>
<li>R/Global.R: Fix a bug where some SNP which are not in reference data (‘snpInfoDir’) were not extended with SNP in LD.</li>
<li>R/Global.R: Fix bug for the extend method ‘LD block’ where SNPs which are not in linkage with another SNP, are removed.</li>
<li>man/*.Rd: Update Rd files with the R code correction.</li>
</ul>
</div>
    <div id="snpenrichment-102" class="section level1">
<h1 class="page-header" data-toc-text="1.0.2">
<a href="#snpenrichment-102" class="anchor"></a>snpEnrichment 1.0.2<small> Unreleased </small>
</h1>
<ul>
<li>R/Global.R: Matrix or data.frame are not anymore evaluated in the slot ‘Parameters’ of Enrichment object.</li>
<li>R/enrichment.R: Fix errors that occurred when the SNP list is too low in ‘reSample’ function.</li>
<li>R/enrichment.R: Fix an error that occured in ‘compareEnrichment’, when non-resampled Enrichment objects are given in the list fields.</li>
</ul>
</div>
    <div id="snpenrichment-101" class="section level1">
<h1 class="page-header" data-toc-text="1.0.1">
<a href="#snpenrichment-101" class="anchor"></a>snpEnrichment 1.0.1<small> Unreleased </small>
</h1>
<ul>
<li>R/Global.R: Some minor code optimization and simplification.</li>
<li>R/enrichment.R: The method ‘summary’ is not anymore masked by loading ‘snpEnrichment’ package.</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://mickael.canouil.fr'>Mickaël Canouil</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


